Packs/Netskope/Integrations/NetskopeEventCollector/NetskopeEventCollector.py

```diff
@@ -101,77 +101,123 @@ def populate_modeling_rule_fields(event: dict, event_type: str):
-        et_limit = limit
-        while len(et_events) < limit:
-            page_limit = min(et_limit, MAX_EVENTS_PAGE_SIZE)
-            if event_type == 'alert':
-                response = client.get_alerts_request_v1(last_run, page_limit)
-            else:
-                response = client.get_events_request_v1(event_type, last_run, page_limit)
+        next_batch = True
+        events = []
+        skip = 0
+        while next_batch:
+            if api_version == 'v1':
+                if event_type == 'alert':
+                    response = client.get_alerts_request_v1(last_run, skip, limit, is_command)
+                else:
+                    response = client.get_events_request_v1(event_type, last_run, skip, limit, is_command)
 
-            if response.get('status') != 'success' or not (results := response.get('data', [])):  # type: ignore
-                break
+                if response.get('status') != 'success':  # type: ignore
+                    break
 
-            events = []
-            for event in results:
-                if event.get('timestamp') == last_run[event_type] and event.get('_id') not in last_run_ids:
-                    events.append(event)
-                    last_run_ids.add(event['_id'])
-                else:
-                    last_run[f'{event_type}-ids'] = [event['_id']]
-                    last_run_ids = set(event['_id'])
-                    last_run[event_type] = event['timestamp']
-                    events.append(event)
-
-            et_events.extend(events)
-            # prepare for the next iteration
-            last_run[f'{event_type}-ids'] = list(last_run_ids)
-            demisto.debug(f'Initialize last run again - {last_run}')
-            et_limit -= len(events)
-            # results were less than a page - continue fetching next time
-            if len(results) < MAX_EVENTS_PAGE_SIZE:
+                results = response.get('data', [])  # type: ignore
+
+            else:  # API version == v2
+                response = client.get_events_request_v2(event_type, last_run, skip, limit, is_command)
+                if response.get('ok') != 1:
+                    break
+
+                results = response.get('result', [])
+
+            demisto.debug(f'The number of events pagination - {len(results)}')
+            events.extend(results)
+            if len(results) == MAX_EVENTS_PAGE_SIZE:
+                skip += MAX_EVENTS_PAGE_SIZE
+
+            if len(results) < MAX_EVENTS_PAGE_SIZE or not results or len(events) == MAX_SKIP:
+                # This means that we either finished going over all results or that we have reached the
+                # limit of accumulated events.
                 break
 
+        final_events, new_last_run, last_run_ids = dedup_by_id(last_run, events, event_type, limit)
+        et_events.extend(final_events)
+        # prepare for the next iteration
+        last_run[f'{event_type}-ids'] = list(last_run_ids)
```

Comment:
Per my comment in the dedup function, I don't think we have any need to work with the last run here. Also seems that we have ignored the `new_last_run`, let's just return it back in the call stack until the part we set the last run it will add more clarity and will be more bug-proof.
```suggestion
        # prepare for the next iteration
        new_last_run.update(partial_last_run)
```